File: 1727.largest-submatrix-with-rearrangements.py
#
# @lc app=leetcode id=1727 lang=python3
#
# [1727] Largest Submatrix With Rearrangements
#

# @lc code=start
from typing import List


class Solution:
    '''
    This is essentially Leetcode 85 Maximal Rectangle. Each time we find the maximum rectangle in histogram for each row, we know we will get the largest rectangle if we can sort the counts 
    '''
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights = [0] + heights + [0]
        stack = []
        max_area = 0
        for i in range(len(heights)):
            index = i
            while stack and stack[-1][1] > heights[i]:
                index, height = stack.pop()
                area = height*(i - index)
                max_area = max(max_area, area)
            stack.append([index, heights[i]])        
        return max_area
    
    def largestSubmatrix(self, matrix: List[List[int]]) -> int:
        max_area, m, n = 0, len(matrix), len(matrix[0])
        row = [0] * n
        for i in range(m):
            for j in range(n):
                row[j] = row[j] + 1 if matrix[i][j] else 0
            max_area = max(max_area, self.largestRectangleArea(sorted(row)))
        return max_area
# @lc code=end

File: 1727.largest-submatrix-with-rearrangements.cpp
/*
 * @lc app=leetcode id=1727 lang=cpp
 *
 * [1727] Largest Submatrix With Rearrangements
 */

// @lc code=start
#include <vector>
#include <algorithm>

class Solution {
public:
    int largestRectangleArea(std::vector<int>& heights) {
        int n = heights.size();
        std::vector<int> h = {0};
        h.insert(h.end(), heights.begin(), heights.end());
        h.push_back(0);
        
        std::vector<std::pair<int, int>> stack;
        int maxArea = 0;
        
        for (int i = 0; i < h.size(); ++i) {
            int index = i;
            while (!stack.empty() && stack.back().second > h[i]) {
                auto [ind, height] = stack.back();
                stack.pop_back();
                int area = height * (i - ind);
                maxArea = std::max(maxArea, area);
            }
            stack.emplace_back(index, h[i]);
        }
        
        return maxArea;
    }
    
    int largestSubmatrix(std::vector<std::vector<int>>& matrix) {
        int maxArea = 0;
        int m = matrix.size();
        int n = matrix[0].size();
        
        std::vector<int> row(n, 0);
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                row[j] = matrix[i][j] ? row[j] + 1 : 0;
            }
            
            std::vector<int> sortedRow(row);
            std::sort(sortedRow.begin(), sortedRow.end());
            
            maxArea = std::max(maxArea, largestRectangleArea(sortedRow));
        }
        
        return maxArea;
    }
};

// @lc code=end

